packages/fxa-payments-server/src/routes/Subscriptions/PaymentUpdateFormV2.test.tsx

```diff
@@ -0,0 +1,679 @@
+    expect(screen.queryByTestId('paymentForm')).toBeInTheDocument();
+  });
+
+  async function commonSubmitSetup({
+    apiClientOverrides = defaultApiClientOverrides(),
+    stripeOverride = defaultStripeOverride(),
+    refreshSubscriptions = jest.fn(),
+    ...props
+  } = {}) {
+    render(
+      <Subject
+        {...{
+          apiClientOverrides,
+          stripeOverride,
+          refreshSubscriptions,
+          ...props,
+        }}
+      />
+    );
+
+    expect(screen.queryByTestId('payment-update')).toBeInTheDocument();
+    fireEvent.click(screen.getByTestId('change-button'));
+
+    await act(async () => {
+      mockStripeElementOnChangeFns.cardElement(
+        elementChangeResponse({ complete: true, value: 'test' })
+      );
+    });
+
+    fireEvent.change(screen.getByTestId('name'), {
+      target: { value: DISPLAY_NAME },
+    });
+    fireEvent.blur(screen.getByTestId('name'));
+
+    await act(async () => {
+      fireEvent.click(screen.getByTestId('submit'));
+    });
+
+    return {
+      apiClientOverrides,
+      stripeOverride,
+      refreshSubscriptions,
+    };
+  }
+
+  const EXPECTED_CONFIRM_CARD_SETUP_ARGS = [
+    FILTERED_SETUP_INTENT.client_secret,
+    {
+      payment_method: {
+        card: { isMockElement: true, testid: 'cardElement' },
+        billing_details: { name: DISPLAY_NAME },
+      },
+    },
+  ];
+
+  it('updates payment method as expected', async () => {
+    const {
+      apiClientOverrides,
+      stripeOverride,
+      refreshSubscriptions,
+    } = await commonSubmitSetup();
+
+    expect(
+      screen.queryByTestId('error-message-container')
+    ).not.toBeInTheDocument();
+    expect(screen.queryByTestId('paymentForm')).not.toBeInTheDocument();
+    expect(apiClientOverrides.apiCreateSetupIntent).toHaveBeenCalledTimes(1);
+    expect(stripeOverride.confirmCardSetup).toHaveBeenCalledWith(
+      ...EXPECTED_CONFIRM_CARD_SETUP_ARGS
+    );
+    expect(
+      apiClientOverrides.apiUpdateDefaultPaymentMethod
+    ).toHaveBeenCalledWith({
+      paymentMethodId: CONFIRM_CARD_SETUP_RESULT.setupIntent.payment_method,
+    });
+    expect(refreshSubscriptions).toBeCalledTimes(1);
+  });
+
+  it('clears the displayed error if the form changes', async () => {
+    await commonSubmitSetup({
+      apiClientOverrides: {
+        ...defaultApiClientOverrides(),
+        apiCreateSetupIntent: jest
+          .fn()
+          .mockRejectedValue('barf apiCreateSetupIntent'),
+      },
+    });
+
+    expect(screen.queryByTestId('error-message-container')).toBeInTheDocument();
+
+    await act(async () => {
+      fireEvent.change(screen.getByTestId('name'), {
+        target: { value: 'Bar Bar' },
+      });
+    });
+
+    await waitForExpect(() =>
+      expect(
+        screen.queryByTestId('error-message-container')
+      ).not.toBeInTheDocument()
+    );
+  });
+
+  it('displays an error if apiCreateSetupIntent throws', async () => {
```

Comment:
Was trying to find a way to DRY up and make these tests less verbose. They basically each just test one failure condition and the expected sequence of API calls & UI changes. Gave up and just did copypasta with tweaks.